src/throttle/core.clj

```diff
@@ -61,6 +61,45 @@
                                :errors      {exception-field-key message}}))))
   (swap! attempts conj [keyy (System/currentTimeMillis)]))
 
+(defn- remaining-attempts
+  [^Throttler {:keys [attempts attempts-threshold]} keyy]
+  (let [[[_ most-recent-attempt-ms], :as keyy-attempts] (filter (fn [[k _]] (= k keyy)) @attempts)]
+    (if most-recent-attempt-ms
+      (let [num-recent-attempts (count keyy-attempts)]
+        (- attempts-threshold num-recent-attempts))
+      attempts-threshold)))
+
+(defn- make-throttling-ex
+  [{:keys [exception-field-key] :as throttler} keyy]
+  (let [delay-ms (calculate-delay throttler keyy)
+        message  (if delay-ms
+                   (format "Too many attempts! You must wait %d seconds before trying again."
+                           (int (math/round (/ delay-ms 1000))))
+                   "Too many attempts! Please try again later.")]
+    (ex-info message {:errors {exception-field-key message}})))
+
+(defn- add-attempt!
+  [^Throttler {:keys [attempts]} keyy]
+  (swap! attempts conj [keyy (System/currentTimeMillis)]))
+
+(defn do-with-throttling
+  [^Throttler throttler keyy f]
+  (remove-old-attempts throttler)
+  (try
+    (when (<= (remaining-attempts throttler keyy) 0)
+      (throw (make-throttling-ex throttler keyy)))
+    (f)
+    (catch Throwable e
+      (add-attempt! throttler keyy)
+      (throw e))))
+
+(defmacro with-throttling {:style/indent 2}
+  [throttler keyy & body]
+  "Do BODY if failed attempts for KEYY on THROTTLER has not been exceeded.
+  If BODY throws an exception, a failed attempt is counted and the exception is re-thrown. If the failed attempts
+  threshold is exceeded, an exception is thrown and BODY is not executed. Attempts made while the threshold is
+  exceeded are counted as additional failed attempts."
+  `(do-with-throttling ~throttler ~keyy (fn [] ~@body)))
```

Comment:
Some of the really old code I wrote has arg names in `UPPER-CASE`. That's the convention in Emacs Lisp and Common Lisp and at the time AFAIK there wasn't any settled convention in Clojure, which was why I was doing stuff like that. (I also wrote a Emacs minor mode to highlight the uppercase words in docstrings which made things nice to look at.)

These days pretty much everyone has settled on using backticks for args, it's even in the Style Guide now: https://guide.clojure.style/#document-pos-arguments

So if you see any code that is like `ARG` feel free to change it. I've been trying to change as I come across it. But definitely don't add any new docstrings like that.

Oh yeah also docstrings go before arglists, altho the linter should be catching that for you

Reply:
Yeah, I noticed that, but opted for consistency. I'll update them all :+1: